core/classifiers/classifier_utils.py

```diff
@@ -60,15 +59,101 @@ def extract_svm_parameters(clf):
     }
 
 
-def unicode_validator_for_classifier_data(var):
-    """Validates that incoming object contains unicode literal strings."""
-    if isinstance(var, dict):
-        for k in var.keys():
+def unicode_validator_for_classifier_data(classifier_data):
+    """Validates that incoming object contains unicode literal strings.
+
+    Args:
+        classifier_data: *. The trained classifier model data.
+
+    Raises:
+        Exception. If any of the strings in classifier data is not a unicode
+            string.
+    """
+    if isinstance(classifier_data, dict):
+        for k in classifier_data.keys():
             if isinstance(k, str):
                 raise Exception('Expected %s to be unicode but found str.' % k)
-            unicode_validator_for_classifier_data(var[k])
-    if isinstance(var, (list, set, tuple)):
-        for item in var:
+            unicode_validator_for_classifier_data(classifier_data[k])
+    if isinstance(classifier_data, list):
+        for item in classifier_data:
             unicode_validator_for_classifier_data(item)
-    if isinstance(var, str):
-        raise Exception('Expected \'%s\' to be unicode but found str.' % var)
+    if isinstance(classifier_data, str):
+        raise Exception(
+            'Expected \'%s\' to be unicode but found str.' % classifier_data)
+
+
+FLOAT_INDICATOR_KEY = 'float_values'
+
+
+# pylint: disable=too-many-branches
+def convert_float_numbers_to_string_in_classifier_data(classifier_data):
+    """Converts all floating point numbers in classifier data to string.
+
+    Args:
+        classifier_data: dict|list. The original classifier data which needs
+            conversion of floats to strings.
+
+    Returns:
+        Dict|List. Modified classifier data in which float values are converted
+            into strings and each dict/subdict is augmented with a special key
+            'float_values' which contains list of keys whose values have
+            undergone the transformation.
+    """
+    if isinstance(classifier_data, dict):
+        if FLOAT_INDICATOR_KEY in classifier_data:
+            raise Exception(
+                'Classifier data already contains a %s key' %
+                FLOAT_INDICATOR_KEY)
+        float_fields = []
+        for k in classifier_data:
+            if isinstance(classifier_data[k], (basestring, int)):
+                classifier_data[k] = classifier_data[k]
+            elif isinstance(classifier_data[k], dict):
+                classifier_data[k] = (
+                    convert_float_numbers_to_string_in_classifier_data(
+                        classifier_data[k]))
+            elif isinstance(classifier_data[k], list):
+                new_list, is_any_value_float = (
```

Comment:
This seems a bit off, design-wise -- it looks like your return value is quite different in the two cases. (In one it doesn't come with a second boolean but in the other it does.)

Will you have a case where the list contains dicts rather than just raw floats? If not, I suggest just accepting the latter case and handling it right here in this function itself -- no need to recurse. We can always update to support the former case later if need be. That would also simplify things since this function would always receive a dict.

Reply:
Yep, it is possible to have dicts within list and I think CodeClassifier does have one parameter with such structure and recursion is required as we also have sublists within a list. So I can think of two approaches here:
1. Do an equality based check. That is check if the old_list and new_list are equal or not. If they are not then we know that something is modified and so we need to handle this list during decoding. I tried this approach and it works. I think this is what we should go with.
2. Another approach is to replace every list with a dict with the following structure:
    ```
    {
        'values': [transformed list],
        'float_values': 'values'
    }
    ```
But this can make decoding algorithm complex because now we have to handle dicts in two different ways: a normal dict and a transformed list dict.

Reply:
Hm, interesting -- in the code classifier data there are even numbers represented as strings. (See the keys of fingerprint_data.)

I do see lists within lists, but I don't see dicts within lists. If that's the case maybe we can say that if a list key is in float_values then every element in that list and its sublists needs to be transformed?

Reply:
That way, you can have a self-contained recursive helper method that takes a list and converts everything in it to a float.

So then the top-level requirements that we would impose on the classifier data structure are:
- only ints, floats, dicts, lists
- lists contain only lists, floats, strings
- in any given list, all the leaf-values are floats, or all the leaf-values are strings (no mix-and-match)

Would this be OK or is it going to be too restrictive?

Reply:
> Hm, interesting -- in the code classifier data there are even numbers represented as strings. (See the keys of fingerprint_data.)

Actually keys are always represented as strings in JSON, even if they are actually integers. So they are actually used as integers but its just json thaat stores them as strings. Code classifier prediction service transforms it back to integers, later.
```
>>> s = {1: 2}
>>> json.dumps(s)
'{"1": 2}'
```

> I do see lists within lists, but I don't see dicts within lists. If that's the case maybe we can say that if a list key is in float_values then every element in that list and its sublists needs to be transformed?

That's exactly what I have done here. The boolean `is_any_value_float` checks that if there is any value is a float and returns it. Based on the return value the key is added. That's why list returns not only list but also a boolean. So that we can know whether we should add this key or not.

> That way, you can have a self-contained recursive helper method that takes a list and converts everything in it to a float.

If you look at the decoder in Oppia then you would find that, that is exactly what decoder does. Once it gets a list it tries to convert everything. It no longer looks for `float_values` when the incoming data is a list.

> * keys of dicts are always strings

1st restriction is already applied by `unicode_validator_for_classifier_data` function.

> * values of dicts must be ints, floats, dicts, lists

 In 2nd, value of the dict must be ints, floats, dicts, lists or **strings**.

> * lists contain only lists, floats, strings

I think we should allow list to contains dicts. Our training data has a list of dicts and, since, we have to recurs for sublists anyway, I don't see why we can't do same for subdicts.

> * in any given list, all the leaf-values are floats, or all the leaf-values are strings (no mix-and-match)

If we allow lists to contain dicts then notion of leaf-values become vague and this is also not necessary. We gain nothing by adding restriction and lose nothing by not adding it. I just don't see why such restriction would be required.

My thinking is to apply only those restrictions which are absolutely necessary and keep things as flexible as possible. That way we can allow a variety of classifiers in Oppia.

Reply:
Ah, so, I think by is_any_value_float I was thinking more like is_every_value_float. And then listing the key of the list in float_values is basically a signal that every value is a float.

The reasoning for every/none is that it's hard to identify specific values in a list, so it would be simpler if we knew that everything is a float vs none is.

I guess lists containing dicts is OK though, as long as the above restriction holds? I.e. in a form that's something like this:

- All lists in classifier_data must either:
  - only contain lists, whose non-dict/list subentries must all have the same type (either float, int or string)
  - only contain dicts, and each dict in the list must have the same key
  - only contain primitive values, and each primitive value must be the same type

Then, for case 2, we can store float keys of subvalues in the subdict, like we do otherwise. And for case 1 or 3, we can store the list key itself as the float key if all non-dict/list subentries of the entire list consist of floats.

Btw, I'm thinking through this sort of approach because I don't really like the two approaches mentioned. But I just realized that, although I don't like (2) for the reasons you stated, I don't think I understand (1). What exactly do you mean by "old list and new list"? How would you decode it? Remember that the point is that there may be strings in the list (maybe from training data) that could accidentally decode to floats, so we need to be more strict than just saying "something has changed".

Reply:
> only contain lists, whose non-dict/list subentries must all have the same type (either float, int or string)

So can we have a structure like `[[1, 2, 3], [[0.4, 0.5, 0.6]]']`? The restrictions are vague. Do you mean to say that every list in its local contex should have same type of values or every list in global context should have same type of values?

It is infeasible from the implementation point of view to check that not only leaf values but also intermediate values have the same type. So, if the above structure is valid then I think it may be possible to have such an implementation.

> What exactly do you mean by "old list and new list"? How would you decode it? Remember that the point is that there may be strings in the list (maybe from training data) that could accidentally decode to floats, so we need to be more strict than just saying "something has changed".

Decoding will be same as now. The comparison is to identify whether this list contained any float values or not. When I say "something has changed" doesn't it mean that it must be due to float-to-string conversion. The algorithm does not modify any other values? So when lists are not equal and something has changed, it implicitly means that some conversion has happened.

Reply:
I think it's more "all floats" or "all non-floats", once you get to the sub-elements of a list. So I think `[[1.0, 2.0, 3.0], [[[0.4, 0.5, 0.6]]]` is fine and `[[1, 2, 3], [[0.4, 0.5, 0.6]]]` is fine if we don't care about treating ints as floats. But `[[1.0, 2.0, 3.0], [['ab', 'bc', 'de']]]` is not fine.

Re your last paragraph, let's say that the input is `[['1', '2', 'a'], [[0.4, 0.5, 0.6]]]`. How will the decoding algorithm know whether it should convert to `[[1, 2, 'a'], [[0.4, 0.5, 0.6]]]` or `[['1', '2', 'a'], [[0.4, 0.5, 0.6]]]`?

Basically -- is there any general specification at all for what the structure and contents of classifier_data could be? It seems to me like currently the answer is "any JSON dict", although I assume we can also say that the elements in any sub-list have the same type. I guess we can't say more than that though, and it's also infeasible to tweak the data into a format where we can say more than that? But then I can't see how you'd get over the issue of it being ambiguous whether a string should be decoded into a string or a float.

Reply:
> I think it's more "all floats" or "all non-floats", once you get to the sub-elements of a list. So I think [[1.0, 2.0, 3.0], [[[0.4, 0.5, 0.6]]] is fine and [[1, 2, 3], [[0.4, 0.5, 0.6]]] is fine if we don't care about treating ints as floats. But [[1.0, 2.0, 3.0], [['ab', 'bc', 'de']]] is not fine.

That requires global context-aware approach. And when you put the dicts into this, it becomes more and more complex. The whole thing just becomes so un-understandable and big mess, everything needs a special case. Even if I build an algorithm to do this, I, myself, wouldn't understand it after a week.

> Re your last paragraph, let's say that the input is [['1', '2', 'a'], [[0.4, 0.5, 0.6]]]. How will the decoding algorithm know whether it should convert to [[1, 2, 'a'], [[0.4, 0.5, 0.6]]] or [['1', '2', 'a'], [[0.4, 0.5, 0.6]]]?

Right now, I am just using a `try ... except` block to see if it is possible to convert a string into float. We can also add regex, so that we don't need a try ... except block. Regex can make sure that content of string is indeed a float value and not int. The only problem then we would have is that, if the float value is actually supposed to be stored as a string. But then that case will be handled by `float_values` key. If it is supposed to be string then `float_values` wouldn't have corresponding keys. Now the only thing that can be problematic is a list like `['0.123', 0.123]`, a list containing a string and a float value. However, the possibility of storing a float value as a string is nearly non-existent. Then we also have `float_values` to make sure that we convert only when we should be. So I think it should be fine 99.99% and when we actually come across the case where a list stores float value as a string and another float value as float, then we can tackle it at that time. The difference will be visible in the data obtained in frontend prediction service. So there is no chance of the error going to production and pass end to end tests successfully.

So, I think, the question is imposing such restrictions really what we need versus the complexity of having such restrictions? As you can see from my argument, there is one specific case that might lead to problems and I am not sure why someone would store a float as a string and not as an actual float.

Reply:
> Basically -- is there any general specification at all for what the structure and contents of classifier_data could be? It seems to me like currently the answer is "any JSON dict", although I assume we can also say that the elements in any sub-list have the same type. I guess we can't say more than that though, and it's also infeasible to tweak the data into a format where we can say more than that? But then I can't see how you'd get over the issue of it being ambiguous whether a string should be decoded into a string or a float.

There is no general specification and it can never be provided. The trained classifier data is driven by prediction service. The only thing that can verify the structure of generated dict is classifier on Oppia-ml and prediction service on Oppia. It directly links them.

> But then I can't see how you'd get over the issue of it being ambiguous whether a string should be decoded into a string or a float.

We can get over it if we just assume that floats will never be actually required as strings and will always be required as floats.

Reply:
OK, I think your point is fair. Can we then document somewhere what the actual specification is for the classifier data, then? I.e. something along the lines of "This data should never contain strings that would pass a float(x) conversion."

And also explain how we would ensure that, if that ambiguity occurs in practice, we will actually catch it before it gets passed around? Otherwise we could end up with lots of corrupted data in our datastore and we wouldn't know. Maybe we can do a check for it prior to sending it back from oppia-ml to oppia -- i.e. ensure that no string within the whole thing can be interpreted as a float.

As long as we have strong safeguards, then I am fine with what you suggest. I agree that the explicit approach is starting to get a bit out of hand.

Reply:
> OK, I think your point is fair. Can we then document somewhere what the actual specification is for the classifier data, then? I.e. something along the lines of "This data should never contain strings that would pass a float(x) conversion."

I agree. Even though we don't know what classifier_data structure should be, we should outline the specific requirements that it must satisfy. Like ideally we would want all the contents in list to have same structure and all primitive values of same type.

> And also explain how we would ensure that, if that ambiguity occurs in practice, we will actually catch it before it gets passed around? Otherwise we could end up with lots of corrupted data in our datastore and we wouldn't know. Maybe we can do a check for it prior to sending it back from oppia-ml to oppia -- i.e. ensure that no string within the whole thing can be interpreted as a float.

Yep, that can be done. So it will be caught on Oppia-ml itself.

Reply:
OK, I think we're on the same page! As long as:

- we have a clear definition of the set of things we accept
- we are certain that we can correctly handle any case in this set, and
- there are strong checks that guarantee that any deviation from this set will be caught before it can do any damage

then I'm fine with whatever solution we use. Sorry for leading you on a bit of a goose chase!

Reply:
I just thought of an approach which would make verifying substructures of list easy. It can be done by BFS (I just thought of an approach that would work) type approach. It will be a separate function since it happens recursively and not iteratively. But it can be done. And now we can be sure that lists have same substructure across all items. Then we won't have the case of one value in list being string and other being float. So we don't need a check whether string is a float or not because each value will either be a string or float. And if it is the string then the corresponding key won't be added in `float_values`.  Actually, I think it makes our algorithm completely deterministic. I will push new implementation, ASAP.